pkg/asset/cluster/tfvars.go

```diff
@@ -417,18 +416,45 @@ func (t *TerraformVariables) Generate(parents asset.Parents) error {
 		for i, w := range workers {
 			workerConfigs[i] = w.Spec.Template.Spec.ProviderSpec.Value.Object.(*machinev1beta1.GCPMachineProviderSpec)
 		}
-		if installConfig.Config.Publish == types.ExternalPublishingStrategy {
-			publicZone, err := gcpconfig.GetPublicZone(ctx, installConfig.Config.GCP.ProjectID, installConfig.Config.BaseDomain)
+		preexistingnetwork := installConfig.Config.GCP.Network != ""
+
+		createFirewallRules := true
+		if installConfig.Config.GCP.CreateFirewallRules == gcp.CreateFirewallRulesDisabled {
+			createFirewallRules = false
+		}
+
+		// Setup defaults for public dns zone
+		createPublicZoneRecords := true
+		publicZoneName := ""
+		publicZoneProject := installConfig.Config.GCP.ProjectID
+		if installConfig.Config.GCP.PublicDNSZone != nil && installConfig.Config.GCP.PublicDNSZone.ProjectID != "" {
+			publicZoneProject = installConfig.Config.GCP.PublicDNSZone.ProjectID
+		}
+
+		switch {
+		case installConfig.Config.Publish != types.ExternalPublishingStrategy:
+			// Do not create public records when not publishing externally.
+			createPublicZoneRecords = false
+		case installConfig.Config.GCP.PublicDNSZone != nil && installConfig.Config.GCP.PublicDNSZone.ID != "":
+			publicZoneName = installConfig.Config.GCP.PublicDNSZone.ID
+		default:
+			// Search the project for a dns zone with the specified base domain.
+			publicZone, err := gcpconfig.GetPublicZone(ctx, publicZoneProject, installConfig.Config.BaseDomain)
 			if err != nil {
 				return errors.Wrapf(err, "failed to get GCP public zone")
 			}
 			publicZoneName = publicZone.Name
 		}
-		preexistingnetwork := installConfig.Config.GCP.Network != ""
 
-		createFirewallRules := true
-		if installConfig.Config.GCP.CreateFirewallRules == gcp.CreateFirewallRulesDisabled {
-			createFirewallRules = false
+		// Setup defaults for private DNS Zone
+		createPrivateZone := true
+		createPrivateZoneRecords := true
+		privateZoneName := ""
+		privateZoneProject := ""
+		if installConfig.Config.GCP.PrivateDNSZone != nil && installConfig.Config.GCP.PrivateDNSZone.ID != "" {
+			createPrivateZone = false
```

Comment:
This seems counter intuitive to me. When PrivateDNSZone != nil and the ID is provided, we don't create private zones?

Reply:
no that means that one exists already we DO NOT want to create one. We are assuming it is already created and provided by the user. This option of adding the privateDNSZone.ID is not to name a new private zone but to use an existing one.

Reply:
Ok. Got it. I think I was equating createPrivateZone with what createPrivateZoneRecords is doing.